### PR TITLE
Add discovery session

### PR DESCRIPTION
**Pull request by dati91**

### CODE REVIEW

**Comment by dati91** [comment]
@mmatyas r?

File: components/net/bluetooth_thread.rs [hunk 1]
```diff
@@ -25,6 +28,7 @@ const PRIMARY_SERVICE_ERROR: &'static str = "No primary service found";
 const CHARACTERISTIC_ERROR: &'static str = "No characteristic found";
 const DESCRIPTOR_ERROR: &'static str = "No descriptor found";
 const VALUE_ERROR: &'static str = "No characteristic or descriptor found with that id";
+const MAX_DISCOVERY_TRY: u8 = 5;
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: `MAX_DISCOVERY_ATTEMPTS` maybe?

File: components/net/bluetooth_thread.rs [hunk 2]
```diff
@@ -355,30 +359,43 @@ impl BluetoothManager {
...
-                                     vendor_id_source: device.get_vendor_id_source().ok(),
-                                     vendor_id: device.get_vendor_id().ok(),
-                                     product_id: device.get_product_id().ok(),
-                                     product_version: device.get_device_id().ok(),
-                                     appearance: device.get_appearance().ok(),
-                                     tx_power: device.get_tx_power().ok().map(|p| p as i8),
-                                     rssi: device.get_rssi().ok().map(|p| p as i8),
-                                 });
-                return drop(sender.send(message));
+        while discovery_try != 0 {
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: wouldn't a simple `while discovery_try {` work?

Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: Apparently not. In this case, how about `discovery_try > 0`?

**Comment by dati91** [comment]
@mmatyas r?

File: components/net/bluetooth_thread.rs [hunk 3]
```diff
@@ -355,30 +360,43 @@ impl BluetoothManager {
...
         };
-        let devices = self.get_and_cache_devices(&mut adapter);
-        if devices.is_empty() {
-            return drop(sender.send(Err(String::from(DEVICE_ERROR))));
+        let mut discovery_attempts = MAX_DISCOVERY_ATTEMPTS;
+        let session = BluetoothDiscoverySession::create_session(adapter.get_object_path()).ok();
+        if let Some(ref s) = session {
+            let _ = s.start_discovery();
+        } else {
+            discovery_attempts = 1;
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: ahogy beszéltük

File: components/net/bluetooth_thread.rs [hunk 4]
```diff
@@ -355,30 +360,43 @@ impl BluetoothManager {
...
-                                     product_id: device.get_product_id().ok(),
-                                     product_version: device.get_device_id().ok(),
-                                     appearance: device.get_appearance().ok(),
-                                     tx_power: device.get_tx_power().ok().map(|p| p as i8),
-                                     rssi: device.get_rssi().ok().map(|p| p as i8),
-                                 });
-                return drop(sender.send(message));
+        while discovery_attempts > 0 {
+            let devices = self.get_and_cache_devices(&mut adapter);
+            let matched_devices: Vec<BluetoothDevice> = devices.into_iter()
```
Comment (zakorgy) [inline components/net/bluetooth_thread.rs]: Can we leave here the type annotation?

File: components/net/bluetooth_thread.rs [hunk 5]
```diff
@@ -203,6 +209,36 @@ impl BluetoothManager {
...
     }
 
+    #[cfg(target_os = "linux")]
+    fn select_device(&mut self, devices: Vec<BluetoothDevice>) -> Option<String> {
+        let mut device_list: Vec<String> = vec!();
+        for device in devices {
+           device_list.push(device.get_address().unwrap_or("".to_string()));
+           device_list.push(device.get_name().unwrap_or("".to_string()));
+        }
+        let device_list_vec: Vec<&str> = device_list.iter()
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: kicsit becsapós ez a `_vec`, mert az előző is Vec volt

File: components/net/bluetooth_thread.rs [hunk 6]
```diff
@@ -355,16 +391,18 @@ impl BluetoothManager {
             Some(a) => a,
             None => return drop(sender.send(Err(String::from(ADAPTER_ERROR)))),
         };
-        let devices = self.get_and_cache_devices(&mut adapter);
-        if devices.is_empty() {
-            return drop(sender.send(Err(String::from(DEVICE_ERROR))));
+        let session = BluetoothDiscoverySession::create_session(adapter.get_object_path()).ok();
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:403]: what if `BluetoothDiscoverySession::create_session(adapter.get_object_path())`  is not ok?

Comment (dati91) [inline components/net/bluetooth_thread.rs:403]: then you won't discover any new device

File: components/net/bluetooth_thread.rs [hunk 7]
```diff
@@ -203,6 +215,36 @@ impl BluetoothManager {
         None
     }
 
+    #[cfg(target_os = "linux")]
+    fn select_device(&mut self, devices: Vec<BluetoothDevice>) -> Option<String> {
+        let mut dialog_rows: Vec<String> = vec!();
+        for device in devices {
+           dialog_rows.extend_from_slice(&[device.get_address().unwrap_or("".to_string()),
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: indent

File: components/net/bluetooth_thread.rs [hunk 8]
```diff
@@ -207,6 +215,36 @@ impl BluetoothManager {
...
+        }
+        let dialog_rows: Vec<&str> = dialog_rows.iter()
+                                                .map(|s| s.as_ref())
+                                                .collect();
+        let dialog_rows: &[&str] = &dialog_rows[..];
+
+        if let Some(device) = tinyfiledialogs::list_dialog(DIALOG_TITLE,
+                                                           &[DIALOG_COLUMN_ID, DIALOG_COLUMN_NAME],
+                                                           Some(dialog_rows)) {
+            return device.rsplit("|").map(|s| s.to_string()).collect::<Vec<String>>().pop();
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:236]: comment?

File: components/net/bluetooth_thread.rs [hunk 9]
```diff
@@ -25,6 +30,13 @@ const PRIMARY_SERVICE_ERROR: &'static str = "No primary service found";
 const CHARACTERISTIC_ERROR: &'static str = "No characteristic found";
 const DESCRIPTOR_ERROR: &'static str = "No descriptor found";
 const VALUE_ERROR: &'static str = "No characteristic or descriptor found with that id";
+const DISCOVERY_TIMEOUT: u64 = 1500;
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:34]: a comment would be nice about why did you choose this value

File: components/net/bluetooth_thread.rs [hunk 10]
```diff
@@ -25,6 +30,13 @@ const PRIMARY_SERVICE_ERROR: &'static str = "No primary service found";
...
+#[cfg(target_os = "linux")]
+const DIALOG_TITLE: &'static str = "Choose a Device";
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: can't we move these DIALOG values to a struct or something?

File: components/net/bluetooth_thread.rs [hunk 11]
```diff
@@ -203,6 +215,36 @@ impl BluetoothManager {
...
+    #[cfg(target_os = "linux")]
+    fn select_device(&mut self, devices: Vec<BluetoothDevice>) -> Option<String> {
+        let mut dialog_rows: Vec<String> = vec!();
+        for device in devices {
+            dialog_rows.extend_from_slice(&[device.get_address().unwrap_or("".to_string()),
+                                            device.get_name().unwrap_or("".to_string())]);
+        }
+        let dialog_rows: Vec<&str> = dialog_rows.iter()
+                                                .map(|s| s.as_ref())
+                                                .collect();
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:228]: can you check if `.collect()[..]` works?,

File: components/net/bluetooth_thread.rs [hunk 12]
```diff
@@ -203,6 +215,36 @@ impl BluetoothManager {
...
+        let dialog_rows: &[&str] = &dialog_rows[..];
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: magic

**Comment by mmatyas** [comment]
looking good, just some minor fötts

File: components/net/bluetooth_thread.rs (hunk 10 above)
Comment (dati91) [inline components/net/bluetooth_thread.rs]: I think it is simpler this way.

File: components/net/bluetooth_thread.rs (hunk 11 above)
Comment (dati91) [inline components/net/bluetooth_thread.rs:228]: It does not work

File: components/net/bluetooth_thread.rs [hunk 13]
```diff
@@ -30,9 +30,10 @@ const PRIMARY_SERVICE_ERROR: &'static str = "No primary service found";
 const CHARACTERISTIC_ERROR: &'static str = "No characteristic found";
 const DESCRIPTOR_ERROR: &'static str = "No descriptor found";
 const VALUE_ERROR: &'static str = "No characteristic or descriptor found with that id";
+// The discovery session needs some time to find any nerby devices
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: nerby? ;)

File: components/net/bluetooth_thread.rs [hunk 14]
```diff
@@ -225,11 +226,13 @@ impl BluetoothManager {
         let dialog_rows: Vec<&str> = dialog_rows.iter()
                                                 .map(|s| s.as_ref())
                                                 .collect();
-        let dialog_rows: &[&str] = &dialog_rows[..];
+        let dialog_rows: &[&str] = dialog_rows.as_slice();
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:229]: :+1:

File: components/net/bluetooth_thread.rs [hunk 15]
```diff
@@ -225,11 +226,13 @@ impl BluetoothManager {
...
 
         if let Some(device) = tinyfiledialogs::list_dialog(DIALOG_TITLE,
                                                            &[DIALOG_COLUMN_ID, DIALOG_COLUMN_NAME],
                                                            Some(dialog_rows)) {
+            // The device string format will be "Address|Name".
+            // We need the first part, therefore we reverse the split and use the last item.
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:235]: nice :+1:

File: components/net/bluetooth_thread.rs [hunk 16]
```diff
@@ -355,26 +400,30 @@ impl BluetoothManager {
...
         }
-
+        let devices = self.get_and_cache_devices(&mut adapter);
         let matched_devices: Vec<BluetoothDevice> = devices.into_iter()
                                                            .filter(|d| matches_filters(d, options.get_filters()))
                                                            .collect();
-        for device in matched_devices {
-            if let Ok(address) = device.get_address() {
+        if let Some(address) = self.select_device(matched_devices) {
+            if let Some(device) = self.get_device(&mut adapter, &*address) {
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: whats this `&*address`

**Pull request merged**